institutions-api/db/db.py

```diff
@@ -0,0 +1,90 @@
+from fastapi import HTTPException
+
+# DB connection based on secrets populated by the crunchydata postgres operator
+engine = create_engine(
+    f'postgresql://{environ["PG_USER"]}:{urllib.parse.quote_plus(environ["PG_PASSWORD"])}@{environ["PG_HOST"]}:{environ["PG_PORT"]}/{environ["PG_DATABASE"]}'
+)
+
+Base.metadata.create_all(engine)
+
+DbSession = sessionmaker(bind=engine)
+
+ROR_ID_TYPE = 'ror_id'
+
+def _ror_id_type(session: Session) -> IdentifierType:
+    """ Get the IdentifierType entity that corresponds to ROR ID """
+    return session.scalars(select(IdentifierType).where(IdentifierType.name == ROR_ID_TYPE)).first()
+
+def _full_osg_id(short_id):
+    """ Get the full osg-htc url of an institution based on its ID suffix """
+    return f"{OSG_ID_PREFIX}{short_id}"
+
+
+def get_institutions() -> List[InstitutionModel]:
+    """ Get a sorted list of every valid institution """
+    with DbSession() as session:
+        institutions = session.scalars(select(Institution)
+            .where(Institution.valid)
+            .order_by(Institution.name)).all()
+        return [InstitutionModel.from_institution(i) for i in institutions]
+
+def get_institution(short_id: str) -> InstitutionModel:
+    """ Get an existing institution by ID """
+    with DbSession() as session:
+        institution = session.scalars(select(Institution)
+                .where(Institution.topology_identifier == _full_osg_id(short_id))).first()
+
+        if institution is None:
+            return HTTPException(404, f"No institution found with id {short_id}")
+
+        return InstitutionModel.from_institution(institution)
+
+def add_institution(institution: InstitutionModel, author: OIDCUserInfo):
+    """ Create a new institution """
+    with DbSession() as session:
+        inst = Institution(institution.name, institution.id, author.id)
+        if institution.ror_id:
+            inst.identifiers = [InstitutionIdentifier(_ror_id_type(session), institution.ror_id)]
+        session.add(inst)
+        session.commit()
+
+def update_institution(short_id: str, institution: InstitutionModel, author: OIDCUserInfo):
+    """ Update an existing institution """
+    with DbSession() as session:
+        to_update = session.scalar(select(Institution)
+            .where(Institution.topology_identifier == _full_osg_id(short_id)))
+
+        if to_update is None:
+            return HTTPException(404, f"No institution found with id {short_id}")
+        
+        to_update.name = institution.name
+
+        # delete any existing ror ids
+        ror_id_type = _ror_id_type(session)
+        session.execute(delete(InstitutionIdentifier)
+            .where(InstitutionIdentifier.institution_id == to_update.id)
+            .where(InstitutionIdentifier.identifier_type_id == ror_id_type.id))
+
+        # re-create the ror id if one is given
+        if institution.ror_id:
+            session.add(InstitutionIdentifier(ror_id_type, institution.ror_id, to_update.id))
```

Comment:
I haven't dug up the call tree/UI but this makes me worry that we'll accidentally remove ROR IDs

Reply:
This workflow should remove the ROR ID from the institution if a form with a null ROR ID is submitted, and update (via delete then recreate) the ROR ID if a non-null value is submitted. Input validation is automatically done by FastAPI via the `@model_validator` annotated method on the Pydantic model class, so the code should error out before starting this database transaction, so the chance of getting the DB into an invalid state here is small https://github.com/osg-htc/topology-institutions-api/pull/1/files#diff-ecedbf460b20f277c0699571a8f7041c665b57ec493538b8a504e0b223a69ce6R24

Reply:
Yeah, I guessed that we'd be protected in other ways but my gut tells me that we should belt + suspenders this, i.e. add some protections to this function to prevent data loss in case something unexpected changes in the frontend.

Reply:
eg. add an explicit "remove ROR ID" endpoint for nulling out the ROR ID, then only allow updates in the PUT endpoint?